File: lab07/identify_poet.py
#!/usr/bin/python
import sys
import re
import glob
import math
from collections import defaultdict

# FUNCTION DEFINITIONS
def findWordFrequencyLog(file, wordToMatch):
    totalWords = findTotal(file)
    totalWordMatches = findWordMatches(file, wordToMatch) + 1
    freq = float(totalWordMatches) / float(totalWords)
    return math.log(freq)

def findTotal(filename):
    fileHandle = open(filename, 'r') # open file for reading
    total = 0
    for line in fileHandle:
        words = re.split(r"[^a-zA-Z]", line)
        for word in words:
            if not word == "":
                total += 1
    fileHandle.close()
    return total

def findWordMatches(filename, wordToMatch):
    total = 0
    fileHandle = open(filename, 'r')
    wordToMatch = wordToMatch.lower()
    for line in fileHandle:
        words = re.split(r"[^a-zA-Z]", line)
        for word in words:
            word = word.lower()
            if word == wordToMatch:
                total += 1
    fileHandle.close()
    return total

# MAIN program
# 2-D hash to hold the log sums for each poem
logSums = defaultdict(dict)
# process all the arguments
argList = sys.argv;
poems = [] # empty list
debug = 0 # debug flag to indicate whether debugging should be turned on
i = 0
for arg in argList:
    if arg == "-d":
        debug = 1
    else:
        poems.append(arg) # assume every other argument is a filename
        i += 1

# print (poems[1:])

# determine who wrote which poem
for poem in poems[1:]:
    for file in sorted(glob.glob("poems/*.txt")):
        # open the file each time it needs to be compared to a file in poems/
        fileHandle = open(poem, 'r')
        author = re.sub(r".txt", "", file)
        author = re.sub(r"_", " ", author)
        author = re.sub(r"poems/", "", author)
        for line in fileHandle:
            words = re.split(r"[^a-zA-Z]", line)
            #print (words)
            for word in words:
                if not word == "":
                    logProb = findWordFrequencyLog(file, word)
                    #print ("word = %s and logProb = %.4f" % (word, logProb))
                    if not author in logSums.get(poem, {}):
                        #print ("First time")
                        logSums[poem][author] = logProb
                    else:
                        #print ("Other time")
                        logSums[poem][author] += logProb

        fileHandle.close()
        if debug == 1:
            print ("%s: log probability of %.1f for %s" % (poem, logSums[poem][author], author))

# print out the results
for poem in sorted(logSums.keys()):
    values = sorted(logSums[poem], key=lambda k: logSums[poem][k], reverse=True)
    print ("%s most resembles the work of %s (log-probability=%.1f)" % (poem, values[0], logSums[poem][values[0]]))

File: lab07/identify_poet.pl
#!/usr/bin/perl -w
$debug = 0;
foreach $arg (@ARGV){
	if ($arg eq "-d"){ # enable debugging
		$debug = 1;
	}

	else{
		push @files, $arg; # add argument ot the list of files to open
	}
}

# loop through all the words in the poem
foreach $poem (@files){
	foreach $file (glob "poems/*.txt"){ # $file = $author
		open P, "<$poem" or die "Unable to open poem file.\n";
		# extract the author from the filename
		$author = $file;
		$author =~ s/poems\///g;
		$author =~ s/\.txt//g;
		$author =~ s/_/ /g;
		# traverse through each line of the input poem
		while ($line = <P>){
			@words = split /[^a-zA-Z]/, $line;
			# traverse through each word of the input poem
			foreach $word (@words){
				# find the log prob of the word in the Poet's poem
				if ($word ne ""){
					$logProbability = logProb($file, $word);
					# print "Added $logProbability to hash table.\n";
					if (defined($logSums{$poem}{$author})){
						# print "Defined value is: $logSums{$poem}{$file}\n";
						$logSums{$poem}{$author} += $logProbability;
					}

					else{
						# print "Assigned a new hash value.\n";
						$logSums{$poem}{$author} = $logProbability;
					}
				}
			}
		}

		if ($debug == 1){
			printf "%s: log_probability of %.1f for %s\n", $poem, $logSums{$poem}{$author}, $author;
		}

		close(P);
	}
}

foreach my $poem (sort keys %logSums){
	# swap a and b around for ascending vs descending
	@allKeys = sort { $logSums{$poem}{$b} <=> $logSums{$poem}{$a} } keys %{$logSums{$poem}};
	$author = $allKeys[0];
	printf "%s most resembles the work of %s (log-probability=%.1f)\n", $poem, $author, $logSums{$poem}{$author};
}

# calculates the log probability of a word in a file
sub logProb{
	my $file = $_[0];
	my $wordToSearch = $_[1];
	my $totalWords = findTotalWords($file);
	my $wordsMatched = findTotalMatches($file, $wordToSearch);
	my $frequency = ($wordsMatched + 1) / $totalWords;
	my $logResult = log($frequency);
	return $logResult;
}

# finds the total number of words in a file
sub findTotalWords{
	my $file = $_[0]; # get the filename from the argument list
	my $total = 0;
	open F, "<$file" or die "Failed to open file\n";
	while ($line = <F>){
		my @words = split /[^a-zA-Z]/, $line;
		foreach $word (@words){
			if ($word ne ""){
				$total++;
			}
		}
	}
	close(F);
	return $total;
	#print "$total words\n";
}

# finds how many times a certain word occurs in a file
sub findTotalMatches{
	my $wordToSearch = lc($_[1]); # get the word from the commandline
	my $filename = $_[0];
	my $total = 0;
	open F, "<$filename" or die "Failed to open file for testing\n";
	while ($line = <F>){
		my @words = split /[^a-zA-Z]/, $line;
		foreach $word (@words){
			if (lc($word) eq $wordToSearch){ # the word was found
				$total++;
			}
		}
	}
	close(F);
	return $total;
}
